src/Cake.NuGet/Install/Extensions/NuGetPackageManagerExtensions.cs

```diff
@@ -0,0 +1,258 @@
+            var allSources = new HashSet<SourceRepository>(localSources, new SourceRepositoryComparer());
+            allSources.AddRange(sources);
+            var gatherContext = new GatherContext()
+            {
+                InstalledPackages = Array.Empty<PackageIdentity>(),
+                PrimaryTargets = primaryPackages,
+                TargetFramework = targetFramework,
+                PrimarySources = allSources.ToList(),
+                AllSources = allSources.ToList(),
+                PackagesFolderSource = packageManager.PackagesFolderSourceRepository,
+                ResolutionContext = resolutionContext,
+                AllowDowngrades = downgradeAllowed,
+                ProjectContext = nuGetProjectContext
+            };
+
+            var availablePackageDependencyInfoWithSourceSet = await ResolverGather.GatherAsync(gatherContext, token);
+
+            if (!availablePackageDependencyInfoWithSourceSet.Any())
+            {
+                throw new InvalidOperationException($"Unable to gather dependency information for package '{packageIdentity}'");
+            }
+
+            // Prune the results down to only what we would allow to be installed
+            // Keep only the target package we are trying to install for that Id
+            var prunedAvailablePackages = PrunePackageTree.RemoveAllVersionsForIdExcept(
+                availablePackageDependencyInfoWithSourceSet,
+                packageIdentity).ToList();
+
+            if (!resolutionContext.IncludePrerelease)
+            {
+                prunedAvailablePackages = PrunePackageTree.PrunePreleaseForStableTargets(
+                    prunedAvailablePackages,
+                    new[] { packageIdentity },
+                    new[] { packageIdentity }).ToList();
+            }
+
+            // Step-2 : Call PackageResolver.Resolve to get new list of installed packages
+            var packageResolverContext = new PackageResolverContext(
+                resolutionContext.DependencyBehavior,
+                new[] { packageIdentity.Id },
+                Enumerable.Empty<string>(),
+                Enumerable.Empty<PackageReference>(),
+                Enumerable.Empty<PackageIdentity>(),
+                prunedAvailablePackages,
+                allSources.Select(s => s.PackageSource),
+                new LoggerAdapter(nuGetProjectContext));
+
+            var packageResolver = new PackageResolver();
+            var packagesToInstall = packageResolver.Resolve(packageResolverContext, token)?.ToList();
+
+            if (packagesToInstall == null || !packagesToInstall.Any())
+            {
+                throw new InvalidOperationException(
+                    $"Unable to resolve dependencies for package '{packageIdentity}' with DependencyBehavior '{resolutionContext.DependencyBehavior}'");
+            }
+
+            // Step-3 : Get the list of nuGetProjectActions to perform, install/uninstall on the nugetproject
```

Comment:
Could we break out step 3 into it's own method somehow? A lot is happening here and the method would probably benefit from being smaller.